common/springUtilities/wupgetLoader.lua

```diff
@@ -0,0 +1,270 @@
+-- returns:  fileName, dirPath
+local function pathParts(fullpath)
+	local _, _, basename = string.find(fullpath, "([^\\/:]*)$")
+	local _, _, path = string.find(fullpath, "(.*[\\/:])[^\\/:]*$")
+	return basename or "", path or ""
+end
+
+---@class WupgetInfo a copy of the table returned by the GetInfo() callin with additional file system and parent/child hierarchy info
+---@field name string name from GetInfo()
+---@field layer number layer from GetInfo()
+---@field desc string description from GetInfo()
+---@field author string author from GetInfo()
+---@field license string license from GetInfo()
+---@field enabled string enabled from GetInfo(), true if the wupget is to be enabled by default
+---@field handler string handler field from GetInfo(), true if the wupget wants access to the widget/gadget handler
+---@field filename string Full path to file from content root including file name
+---@field basename string Name of the file
+---@field path string Path to directory containing the wupget
+---@field childPaths nil | string[] | boolean local paths to children files or directories **OR** boolean true to signal that all other .lua files in this folder are children of this wupget. Result of GetChildPaths() callin.
+---@field parent nil | WupgetInfo parent info
+---@field children nil | WupgetInfo[] child infos
+
+---@param wupget
+---@param filename
+---@return WupgetInfo
+local function extractInfo(wupget, filename)
+	local basename, path = pathParts(filename)
+	local info = {
+		filename = filename,
+		basename = basename,
+		path = path
+	}
+
+	if wupget.GetInfo == nil then
+		info.name = basename
+		info.layer = 0
+	else
+		for k, v in pairs(wupget:GetInfo() or {}) do
+			info[k] = v
+		end
+
+		info.name = info.name or basename
+		info.layer = info.layer or 0
+
+		if wupget.GetChildPaths then
+			info.childPaths = wupget:GetChildPaths(path)
+		end
+	end
+
+	return info;
+end
+
+--- sorts wupgets by layer, then orderList entry, then name
+---
+--- child wupgets are placed after their parent wupgets and sorted amongst themselves
+---@return table[] a new list of sorted wupgets
+local function sortedWupgetList(wupgets, orderList, infoAccessor)
```

Comment:
I think this function can be made faster and easier to understand with a clearer tree abstraction and a depth-first search based sort; I'll try to put together some example code soon to demonstrate. I don't want to block the PR with this though, your algorithm already appears to work as expected.

Reply:
eh, I made this up on the spot when getting things to sort how I wanted in the Widget Selector
then I reused it here

Reply:
I don't know what a depth-first search based sort would look like here
I'm just trying to lean on the built-in sort as much as possible

Reply:
https://github.com/salinecitrine/Beyond-All-Reason/commit/25964c233be5fc6a18b03b064acfba4d68d435d6

This is what I had in mind as an alternative sorting algorithm. I did some benchmarks, and it's about 17% slower for mostly flat widget trees (as it is now), and about twice as fast for trees with a lot of children. Not a requirement, mostly just wanted to try it out.

---

I realized something while writing it out, though: isn't there a conflict between loading widgets in parent/child order, and layer order? what if a widget tree has one widget that wants to be layer=-100, won't it be restricted to loading after its parent?

Reply:
You did performance testing? You are a more diligent developer than I.

I doubt that widget trees will be getting very deep. But who knows what the future holds.

-------

Then the parent should have a lower layer too? That or it shouldn't be a child widget?

hmm...

I now see a case for allowing multiple  "*.main.lua" files in a folder- so they can be organized together but not be forced to have a parent/child relationship

I should also write in more documentation in the example widget, at least putting in a comment that shows returning a list of paths

Reply:
I don't want to go into the complexity of dependency resolution stuff if it can be helped